dawn/src/dawn/Optimizer/PassLocalVarType.cpp

```diff
@@ -91,29 +99,63 @@ class VarTypeFinder : public ast::ASTVisitorForwarding {
     recursivePropagate(accessID);
   }
 
+  iir::LocalVariableType localVarTypeFromLocationType(ast::LocationType locationType) {
+    switch(locationType) {
+    case ast::LocationType::Cells:
+      return iir::LocalVariableType::OnCells;
+    case ast::LocationType::Edges:
+      return iir::LocalVariableType::OnEdges;
+    case ast::LocationType::Vertices:
+      return iir::LocalVariableType::OnVertices;
+    }
+  }
+
+  // Construct a LocalVariableType from a field accessed within an assignment to / declaration of a
+  // local variable.
+  iir::LocalVariableType inferLocalVarTypeFromField(int fieldAccessID) {
+    iir::LocalVariableType type;
+
+    if(sir::dimension_isa<sir::CartesianFieldDimension>(
+           metadata_.getFieldDimensions(fieldAccessID).getHorizontalFieldDimension())) {
+      // Cartesian case
+      type = iir::LocalVariableType::OnIJ;
+    } else {
+      // Unstructured case, convert from location type
+      type =
+          localVarTypeFromLocationType(metadata_.getDenseLocationTypeFromAccessID(fieldAccessID));
+    }
+    return type;
+  }
+  // Records that to write into `destinationVariable` we need to access `accessedVariable`.
+  // `sourceLocation` is where this happens.
+  void recordVariablePair(int accessedVariable, int destinationVariable,
+                          SourceLocation sourceLocation) {
+    // No self-dependencies
+    if(accessedVariable != destinationVariable) {
+
+      // If the set of dependers for this variable is not there, it means the variable hasn't been
+      // declared
+      DAWN_ASSERT_MSG(dependencyMap_.count(accessedVariable),
+                      "Variable accessed before being declared.");
+      // Need to register a dependency. Pair to be added is from dependency (variable accessed in
```

Comment:
if you think this comment is valuable, replace it by a `registerDependency()` function

Reply:
I think if one reads the description of `dependencyMap_`, then the code is already self-explanatory. I'll remove the comment.